src/Tasks/Microsoft.NET.Build.Tasks/GetPackagesToPrune.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Microsoft.ComponentDetection.Detectors.NuGet;
+using NuGet.Frameworks;
+using NuGet.Versioning;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    public class GetPackagesToPrune : TaskBase
+    {
+        [Required]
+        public string TargetFrameworkIdentifier { get; set; }
+
+        [Required]
+        public string TargetFrameworkVersion { get; set; }
+
+        [Required]
+        public ITaskItem[] FrameworkReferences { get; set; }
+
+        [Output]
+        public ITaskItem[] PackagesToPrune { get; set; }
+
+        protected override void ExecuteCore()
+        {
+            var nugetFramework = new NuGetFramework(TargetFrameworkIdentifier, Version.Parse(TargetFrameworkVersion));
+
+            Dictionary<string, NuGetVersion> packagesToPrune = new();
+
+            var frameworkPackages = FrameworkPackages.GetFrameworkPackages(nugetFramework, FrameworkReferences.Select(fr => fr.ItemSpec).ToArray())
+                .SelectMany(packages => packages);
+
+            foreach (var kvp in frameworkPackages)
+            {
+                if (packagesToPrune.TryGetValue(kvp.Key, out NuGetVersion existingVersion))
+                {
+                    if (kvp.Value > existingVersion)
+                    {
+                        packagesToPrune[kvp.Key] = kvp.Value;
+                    }
+                }
+                else
+                {
+                    packagesToPrune[kvp.Key] = kvp.Value;
+                }
+            }
+
+            PackagesToPrune = packagesToPrune.Select(p =>
```

Comment:
I wonder if we'll want a cache of these per framework?

Reply:
I've added a cache in the latest commit, though I'm not sure it's really worth the risk right now, since this is an opt in feature and the tests won't really exercise the caching, so it's pretty much untested.